Objects/setobject.c

```diff
@@ -2748,11 +2748,7 @@ PySet_Contains(PyObject *anyset, PyObject *key)
         return -1;
     }
 
-    int rv;
-    Py_BEGIN_CRITICAL_SECTION(anyset);
-    rv = set_contains_key((PySetObject *)anyset, key);
-    Py_END_CRITICAL_SECTION();
-    return rv;
+    return _PySet_Contains(_PySet_CAST(anyset), key);
```

Comment:
Isn't that just the definition of

```
static int
set_contains(PyObject *self, PyObject *key)
{
    PySetObject *so = _PySet_CAST(self);
    return _PySet_Contains(so, key);
}
```

And thus should this be

```suggestion
    return set_contains(anyset, key);
```

Sorry I'm not super used to touching the c code in CPython, and thanks for your reply on slack

Reply:
Does that also contradict:


> int PySet_Contains(PyObject *anyset, PyObject *key)
>  Part of the Stable ABI.

>  Return 1 if found, 0 if not found, and -1 if an error is encountered. Unlike the Python `__contains__()` method, **this function does not automatically convert unhashable sets into temporary frozensets**. Raise a TypeError if the key is unhashable. Raise SystemError if anyset is not a set, frozenset, or an instance of a subtype.

Emphasis mine.

Reply:
The remark in the documentation is about the `key`, not about the `anyset` argument. The docs could be clarified by changing the text into

> this function does not accept unhashable sets as a `key` argument

The documentation a bit below:

> Raise SystemError if anyset is not a set, frozenset, or an instance of a subtype.

suggests that a `frozenset` is valid input for `PySet_Contains`. (It would have been nicer to call this method `PyAnySet_Contains` then, but I do not think it is worth the effort to change the name)

Reply:
Sure, but `set_contains_lock_held` does convert the _key_ into a frozenset, which `set_contains_key` does not.

```
        PyErr_Clear();
        Py_hash_t hash;
        Py_BEGIN_CRITICAL_SECTION(key);
        hash = frozenset_hash_impl(key);
        Py_END_CRITICAL_SECTION();
        rv = set_contains_entry(so, key, hash);
```



This succeeds on main, not on this PR:

```
static PyObject *
test_set_contains_does_not_convert_unhashable_set(PyObject *self, PyObject *Py_UNUSED(obj))
{
    // Test that PySet_Contains does NOT automatically convert
    // unhashable sets into temporary frozensets.
    // This is different from the Python-level 'in' operator which does
    // perform this conversion via set_contains_lock_held.

    PyObject *outer_set = PySet_New(NULL);
    PyObject *inner_set = PySet_New(NULL);
    PyObject *num = PyLong_FromLong(42);

    // Add an element to inner_set to make it {42}
    PySet_Add(inner_set, num);

    int result = PySet_Contains(outer_set, inner_set);

    if (result < 0) {
        // Expected: should get an error (TypeError)
        if (PyErr_ExceptionMatches(PyExc_TypeError)) {
            PyErr_Clear();
            Py_RETURN_TRUE;  // Test passes
        }
        // Unexpected error type
        return NULL;
    }

    PyErr_SetString(PyExc_AssertionError,
                    "PySet_Contains should have raised TypeError for unhashable set");
    return NULL;
}

```

Reply:
Oh, I missed that, pushed a fix

Reply:
The comment in the documentation is about the `key`, not the `anyset` argument. So `{1,2} in {1,2, frozenset({1,2})}` is valid Python code (which evaluates to True), but the C equivalent with `PySet_Contains` will generate a type error.